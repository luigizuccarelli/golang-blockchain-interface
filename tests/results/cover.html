
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<title>handlers: Go Coverage Report</title>
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go (95.7%)</option>
				
				<option value="file1">lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go (100.0%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package handlers

import (
        "crypto/sha256"
        "encoding/hex"
        "encoding/json"
        "errors"
        "fmt"
        "io/ioutil"
        "net/http"
        "os"
        "strconv"
        "time"

        "github.com/gorilla/mux"
        "lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/connectors"
        "lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/schema"
)

const (
        CONTENTTYPE     string = "Content-Type"
        APPLICATIONJSON string = "application/json"
)

var (
        Blockchain []schema.BlockInterface
)

func GetBlockChainList(w http.ResponseWriter, r *http.Request, conn connectors.Clients) <span class="cov8" title="1">{
        var response *schema.Response
        addHeaders(w, r)
        response = &amp;schema.Response{Name: os.Getenv("NAME"), StatusCode: "200", Status: "OK", Message: "Blockchain list processed succesfully", BlockChain: Blockchain}
        w.WriteHeader(http.StatusOK)
        b, _ := json.MarshalIndent(response, "", "        ")
        conn.Debug("GetBlockChainList response : %s", string(b))
        fmt.Fprintf(w, "%s", string(b))
}</span>

func GetBlockChain(w http.ResponseWriter, r *http.Request, conn connectors.Clients) <span class="cov8" title="1">{
        var response *schema.Response
        var hold []schema.BlockInterface
        vars := mux.Vars(r)
        addHeaders(w, r)
        index, _ := strconv.Atoi(vars["index"])
        hold = append(hold, Blockchain[index])
        response = &amp;schema.Response{Name: os.Getenv("NAME"), StatusCode: "200", Status: "OK", Message: "Blockchain list processed succesfully", BlockChain: hold}
        w.WriteHeader(http.StatusOK)
        b, _ := json.MarshalIndent(response, "", "        ")
        conn.Debug("GetBlockChain response : %s", string(b))
        fmt.Fprintf(w, "%s", string(b))
}</span>

func WriteBlockChain(w http.ResponseWriter, r *http.Request, conn connectors.Clients) <span class="cov8" title="1">{
        var response *schema.Response
        var in schema.InputData
        addHeaders(w, r)
        body, err := ioutil.ReadAll(r.Body)
        if err != nil </span><span class="cov8" title="1">{
                conn.Error("WriteBlockChain %v", err)
                response = &amp;schema.Response{Name: os.Getenv("NAME"), StatusCode: "500", Status: "KO", Message: fmt.Sprintf("Could not read body data %v\n", err), BlockChain: Blockchain}
                w.WriteHeader(http.StatusInternalServerError)
                b, _ := json.MarshalIndent(response, "", "        ")
                fmt.Fprintf(w, "%s", string(b))
                return
        }</span>
        <span class="cov8" title="1">e := json.Unmarshal(body, &amp;in)
        if e != nil </span><span class="cov8" title="1">{
                conn.Error("WriteBlockChain %v", e)
                response = &amp;schema.Response{Name: os.Getenv("NAME"), StatusCode: "500", Status: "KO", Message: fmt.Sprintf("Could not unmarshal json input data %v\n", err), BlockChain: Blockchain}
                w.WriteHeader(http.StatusInternalServerError)
                b, _ := json.MarshalIndent(response, "", "        ")
                fmt.Fprintf(w, "%s", string(b))
                return
        }</span>
        <span class="cov8" title="1">newBlock, e := generateBlock(Blockchain[len(Blockchain)-1], in)
        if e != nil </span><span class="cov8" title="1">{
                conn.Error("WriteBlockChain %v", e)
                response = &amp;schema.Response{Name: os.Getenv("NAME"), StatusCode: "500", Status: "KO", Message: fmt.Sprintf("Could not write to blockchain %v\n", err), BlockChain: Blockchain}
                w.WriteHeader(http.StatusInternalServerError)
                b, _ := json.MarshalIndent(response, "", "        ")
                fmt.Fprintf(w, "%s", string(b))
                return
        }</span>
        <span class="cov8" title="1">if isBlockValid(newBlock, Blockchain[len(Blockchain)-1]) </span><span class="cov8" title="1">{
                newBlockchain := append(Blockchain, newBlock)
                replaceChain(newBlockchain)
        }</span>
        <span class="cov8" title="1">response = &amp;schema.Response{Name: os.Getenv("NAME"), StatusCode: "200", Status: "OK", Message: "Data processed succesfully", BlockChain: Blockchain}
        w.WriteHeader(http.StatusOK)
        b, _ := json.MarshalIndent(response, "", "        ")
        conn.Debug("WriteBlockChain response : %s", string(b))
        fmt.Fprintf(w, "%s", string(b))</span>
}

func Init(w http.ResponseWriter, r *http.Request) <span class="cov8" title="1">{
        if len(Blockchain) == 0 </span><span class="cov8" title="1">{
                t := time.Now()
                genesisBlock := schema.BlockInterface{0, "Genesis", "A", "B", "C", t.String(), "", ""}
                Blockchain = append(Blockchain, genesisBlock)
                fmt.Fprintf(w, "%s", "{ \"created genesis block\" }")
                return
        }</span>
        <span class="cov0" title="0">fmt.Fprintf(w, "%s", "{ \"NOP genesis block created\" }")</span>
}

func generateBlock(oldBlock schema.BlockInterface, in schema.InputData) (schema.BlockInterface, error) <span class="cov8" title="1">{
        var newBlock schema.BlockInterface
        t := time.Now()
        if in.ObjectA == "" || in.ObjectB == "" || in.ObjectC == "" </span><span class="cov8" title="1">{
                return newBlock, errors.New("object blocks can't be empty")
        }</span>
        <span class="cov8" title="1">newBlock.Index = oldBlock.Index + 1
        newBlock.Timestamp = t.String()
        newBlock.MetaInfo = in.MetaInfo
        newBlock.ObjectA = in.ObjectA
        newBlock.ObjectB = in.ObjectB
        newBlock.ObjectC = in.ObjectC
        newBlock.PrevHash = oldBlock.Hash
        newBlock.Hash = calculateHash(newBlock)
        return newBlock, nil</span>
}

func isBlockValid(newBlock, oldBlock schema.BlockInterface) bool <span class="cov8" title="1">{
        if oldBlock.Index+1 != newBlock.Index </span><span class="cov0" title="0">{
                return false
        }</span>
        <span class="cov8" title="1">if oldBlock.Hash != newBlock.PrevHash </span><span class="cov0" title="0">{
                return false
        }</span>
        <span class="cov8" title="1">if calculateHash(newBlock) != newBlock.Hash </span><span class="cov0" title="0">{
                return false
        }</span>
        <span class="cov8" title="1">return true</span>
}

func replaceChain(newBlocks []schema.BlockInterface) <span class="cov8" title="1">{
        if len(newBlocks) &gt; len(Blockchain) </span><span class="cov8" title="1">{
                Blockchain = newBlocks
        }</span>
}

func calculateHash(block schema.BlockInterface) string <span class="cov8" title="1">{
        record := fmt.Sprintf("%d", block.Index) + block.Timestamp + block.ObjectA + block.ObjectB + block.ObjectC + block.PrevHash
        h := sha256.New()
        h.Write([]byte(record))
        hashed := h.Sum(nil)
        return hex.EncodeToString(hashed)
}</span>

func IsAlive(w http.ResponseWriter, r *http.Request) <span class="cov8" title="1">{
        fmt.Fprintf(w, "{ \"version\" : \"1.0.2\" , \"name\": \""+os.Getenv("NAME")+"\" }")
}</span>

// headers (with cors) utility
func addHeaders(w http.ResponseWriter, r *http.Request) <span class="cov8" title="1">{
        w.Header().Set(CONTENTTYPE, APPLICATIONJSON)
        // use this for cors
        w.Header().Set("Access-Control-Allow-Origin", "*")
        w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, PUT, DELETE")
        w.Header().Set("Access-Control-Allow-Headers", "Accept, Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization")
}</span>
</pre>
		
		<pre class="file" id="file1" style="display: none">package validator

import (
        "errors"
        "fmt"
        "os"
        "strconv"
        "strings"

        "github.com/microlib/simple"
)

// checkEnvars - private function, iterates through each item and checks the required field
func checkEnvar(item string, logger *simple.Logger) error <span class="cov8" title="1">{
        name := strings.Split(item, ",")[0]
        required, _ := strconv.ParseBool(strings.Split(item, ",")[1])
        logger.Trace(fmt.Sprintf("name %s : required %t", name, required))
        if os.Getenv(name) == "" </span><span class="cov8" title="1">{
                if required </span><span class="cov8" title="1">{
                        logger.Error(fmt.Sprintf("%s envar is mandatory please set it", name))
                        return errors.New(fmt.Sprintf("%s envar is mandatory please set it", name))
                }</span> else<span class="cov8" title="1"> {
                        logger.Error(fmt.Sprintf("%s envar is empty please set it", name))
                }</span>
        }
        <span class="cov8" title="1">return nil</span>
}

// ValidateEnvars : public call that groups all envar validations
// These envars are set via the openshift template
// Each microservice will obviously have a diffefrent envars so change where needed
func ValidateEnvars(logger *simple.Logger) error <span class="cov8" title="1">{
        items := []string{
                "LOG_LEVEL,false",
                "SERVER_PORT,true",
                "VERSION,true",
        }
        for x, _ := range items </span><span class="cov8" title="1">{
                if err := checkEnvar(items[x], logger); err != nil </span><span class="cov8" title="1">{
                        return err
                }</span>
        }
        <span class="cov8" title="1">return nil</span>
}
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
