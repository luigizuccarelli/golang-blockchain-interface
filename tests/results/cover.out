mode: set
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:14.59,18.27 4 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:26.2,26.12 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:18.27,19.15 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:19.15,22.4 2 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:22.9,24.4 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:32.50,38.26 2 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:43.2,43.12 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:38.26,39.54 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/validator/validate.go:39.54,41.4 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:29.89,37.2 7 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:39.85,51.2 11 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:53.87,58.16 5 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:66.2,67.14 2 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:75.2,76.14 2 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:84.2,84.59 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:88.2,92.33 5 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:58.16,65.3 6 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:67.14,74.3 6 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:76.14,83.3 6 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:84.59,87.3 2 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:95.51,96.26 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:103.2,103.59 1 0
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:96.26,102.3 5 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:106.104,109.62 3 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:112.2,120.22 9 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:109.62,111.3 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:123.66,124.40 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:127.2,127.40 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:130.2,130.46 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:133.2,133.13 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:124.40,126.3 1 0
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:127.40,129.3 1 0
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:130.46,132.3 1 0
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:136.54,137.38 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:137.38,139.3 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:142.56,148.2 5 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:150.54,152.2 1 1
lmzsoftware.com/lzuccarelli/golang-blockchain-interface/pkg/handlers/handlers.go:155.57,161.2 4 1
